File: typescript/sublist/sublist.test.ts
import List from './sublist'

describe('Sublist', () => {
    it('empty lists', () => {
        const listOne = new List()
        const listTwo = new List()
        const expected = 'equal'

        expect(listOne.compare(listTwo)).toEqual(expected)
    })

    it('empty list within non empty list', () => {
        const listOne = new List()
        const listTwo = new List(1, 2, 3)
        const expected = 'sublist'

        expect(listOne.compare(listTwo)).toEqual(expected)
    })

    it('non empty list contains empty list', () => {
        const listOne = new List(1, 2, 3)
        const listTwo = new List()
        const expected = 'superlist'

        expect(listOne.compare(listTwo)).toEqual(expected)
    })

    it('list equals itself', () => {
        const listOne = new List(1, 2, 3)
        const listTwo = new List(1, 2, 3)
        const expected = 'equal'

        expect(listOne.compare(listTwo)).toEqual(expected)
    })

    it('different lists', () => {
        const listOne = new List(1, 2, 3)
        const listTwo = new List(2, 3, 4)
        const expected = 'unequal'

        expect(listOne.compare(listTwo)).toEqual(expected)
    })

    it('false start', () => {
        const listOne = new List(1, 2, 5)
        const listTwo = new List(0, 1, 2, 3, 1, 2, 5, 6)
        const expected = 'sublist'

        expect(listOne.compare(listTwo)).toEqual(expected)
    })

    it('consecutive', () => {
        const listOne = new List(1, 1, 2)
        const listTwo = new List(0, 1, 1, 1, 2, 1, 2)
        const expected = 'sublist'

        expect(listOne.compare(listTwo)).toEqual(expected)
    })

    it('sublist at start', () => {
        const listOne = new List(0, 1, 2)
        const listTwo = new List(0, 1, 2, 3, 4, 5)
        const expected = 'sublist'

        expect(listOne.compare(listTwo)).toEqual(expected)
    })

    it('sublist in middle', () => {
        const listOne = new List(2, 3, 4)
        const listTwo = new List(0, 1, 2, 3, 4, 5)
        const expected = 'sublist'

        expect(listOne.compare(listTwo)).toEqual(expected)
    })

    it('sublist at end', () => {
        const listOne = new List(3, 4, 5)
        const listTwo = new List(0, 1, 2, 3, 4, 5)
        const expected = 'sublist'

        expect(listOne.compare(listTwo)).toEqual(expected)
    })

    it('at start of superlist', () => {
        const listOne = new List(0, 1, 2, 3, 4, 5)
        const listTwo = new List(0, 1, 2)
        const expected = 'superlist'

        expect(listOne.compare(listTwo)).toEqual(expected)
    })

    it('in middle of superlist', () => {
        const listOne = new List(0, 1, 2, 3, 4, 5)
        const listTwo = new List(2, 3)
        const expected = 'superlist'

        expect(listOne.compare(listTwo)).toEqual(expected)
    })

    it('at end of superlist', () => {
        const listOne = new List(0, 1, 2, 3, 4, 5)
        const listTwo = new List(3, 4, 5)
        const expected = 'superlist'

        expect(listOne.compare(listTwo)).toEqual(expected)
    })

    it('first list missing element from second list', () => {
        const listOne = new List(1, 3)
        const listTwo = new List(1, 2, 3)
        const expected = 'unequal'

        expect(listOne.compare(listTwo)).toEqual(expected)
    })

    it('second list missing element from first list', () => {
        const listOne = new List(1, 2, 3)
        const listTwo = new List(1, 3)
        const expected = 'unequal'

        expect(listOne.compare(listTwo)).toEqual(expected)
    })

    it('order matters to a list', () => {
        const listOne = new List(1, 2, 3)
        const listTwo = new List(3, 2, 1)
        const expected = 'unequal'

        expect(listOne.compare(listTwo)).toEqual(expected)
    })

    it('same digits but different numbers', () => {
        const listOne = new List(1, 0, 1)
        const listTwo = new List(10, 1)
        const expected = 'unequal'

        expect(listOne.compare(listTwo)).toEqual(expected)
    })
})

File: scala/sublist/src/test/scala/SublistTest.scala
import org.scalatest.{Matchers, FunSuite}

/** @version 1.1.0 */
class SublistTest extends FunSuite with Matchers {

  test("empty lists") {
    Sublist.sublist(List(), List()) should be(Sublist.Equal)
  }

  test("empty list within non empty list") {
    Sublist.sublist(List(), List(1, 2, 3)) should be(Sublist.Sublist)
  }

  test("non empty list contains empty list") {
    Sublist.sublist(List(1, 2, 3), List()) should be(Sublist.Superlist)
  }

  test("list equals itself") {
    Sublist.sublist(List(1, 2, 3), List(1, 2, 3)) should be(Sublist.Equal)
  }

  test("different lists") {
    Sublist.sublist(List(1, 2, 3), List(2, 3, 4)) should be(Sublist.Unequal)
  }

  test("false start") {
    Sublist.sublist(List(1, 2, 5), List(0, 1, 2, 3, 1, 2, 5, 6)) should be(
      Sublist.Sublist)
  }

  test("consecutive") {
    Sublist.sublist(List(1, 1, 2), List(0, 1, 1, 1, 2, 1, 2)) should be(
      Sublist.Sublist)
  }

  test("sublist at start") {
    Sublist.sublist(List(0, 1, 2), List(0, 1, 2, 3, 4, 5)) should be(
      Sublist.Sublist)
  }

  test("sublist in middle") {
    Sublist.sublist(List(2, 3, 4), List(0, 1, 2, 3, 4, 5)) should be(
      Sublist.Sublist)
  }

  test("sublist at end") {
    Sublist.sublist(List(3, 4, 5), List(0, 1, 2, 3, 4, 5)) should be(
      Sublist.Sublist)
  }

  test("at start of superlist") {
    Sublist.sublist(List(0, 1, 2, 3, 4, 5), List(0, 1, 2)) should be(
      Sublist.Superlist)
  }

  test("in middle of superlist") {
    Sublist.sublist(List(0, 1, 2, 3, 4, 5), List(2, 3)) should be(
      Sublist.Superlist)
  }

  test("at end of superlist") {
    Sublist.sublist(List(0, 1, 2, 3, 4, 5), List(3, 4, 5)) should be(
      Sublist.Superlist)
  }

  test("first list missing element from second list") {
    Sublist.sublist(List(1, 3), List(1, 2, 3)) should be(Sublist.Unequal)
  }

  test("second list missing element from first list") {
    Sublist.sublist(List(1, 2, 3), List(1, 3)) should be(Sublist.Unequal)
  }

  test("order matters to a list") {
    Sublist.sublist(List(1, 2, 3), List(3, 2, 1)) should be(Sublist.Unequal)
  }

  test("same digits but different numbers") {
    Sublist.sublist(List(1, 0, 1), List(10, 1)) should be(Sublist.Unequal)
  }
}
